motoko/passwords/src/passwords/main.mo

```diff
@@ -0,0 +1,121 @@
+import Array "mo:base/Array";
+import Principal "mo:base/Principal";
+import Option "mo:base/Option";
+
+actor {
+
+  // Types
+
+  type Credentials = {
+    username : Text;
+    password : Text;
+  };
+
+  type Account = {
+    id : Nat;
+    username : Text;
+    hash : Text;
+    salt : Text;
+  };
+
+  type Session = {
+    principal : Principal;
+    account_id : Nat;
+  };
+
+
+  // Data Structures
+
+  // Arrays a sub-optimal choice here but they're easy to understand
+  var accounts : [Account] = [];
+  var sessions : [Session] = [];
+  var next_account_id : Nat = 0;
+
+  // Public Functions
+
+  public func signup (credentials : Credentials) {
```

Comment:
I think this might have some indeterminate behavior if different users both try to `signup` with 'identical credentials', i.e. a Credentials with identical `username` and `password.

i.e. I think there are no checks to make sure that the username/password aren't already taken by someone else, and allowing access to that someone else's account. (correct me if I'm wrong. Hard for me to tell if this should happen in `signup` or `login`)